docs/rendering_templates.md

```diff
@@ -0,0 +1,118 @@
+# Rendering BOSH Templates
+
+You can read more about BOSH templates on [bosh.io](https://bosh.io/docs/jobs/#templates).
+
+Rendering happens using `ExtendedJobs`.
+
+An `ExtendedJob` is required for each BOSH Release and Instance Group defined in a Desired Manifest.
+
+Details are omitted in the example below, for brevity:
+
+```yaml
+---
+
+releases:
+- name: release-a
+- name: release-b
+
+instance_groups:
+- name: instance-group-1
+  jobs:
+  - name: job1
+    release: release-a
+  - name: job2
+    release: release-a
+- name: instance-group-2
+  jobs:
+  - name: job1
+    release: release-a
+    properties:
+      foo: "((bar))"
+  - name: job3
+    release: release-b
+
+variables:
+- name: bar
+  type: password
+```
+
+In this case, to render all required templates, 3 `ExtendedJobs` are created:
+- one for  `instance-group-1`, because it only contains jobs from `release-a`
+- two for `instance-group-2`, because it references jobs from both `release-a` and  `release-b`
+
+The `ExtendedJob` runs a ruby process `TODO: reference to the implementation`
+
+## Input & Output
+
+Input is the desired manifest, generated `Secrets` for all referenced variables and information required for discovery of network addresses.
+
+We take advantage of the `ExtendedJob`'s feature to persist output in a `ConfigMap` or `Secret`.
+
+The output of the rendering process is a JSON object that contains all the rendered templates:
+
+The following is and example definition for the `ExtendedJob` that renders templates for `instance-group-2` and `release-a`.
+The release is inferred from the image itself.
+
+```yaml
+---
+---
+apiVersion: fissile.cloudfoundry.org/v1alpha1
+kind: ExtendedJob
+metadata:
+  name: MyExtendedJob
+spec:
+  output:
+    stdout:
+      secretRef: "mynamespace/mydeployment-release-a-instance-group-2"
+      overwrite: true
+      writeOnFailure: false
+  updateOnConfigChange: true
+  template:
+    spec:
+      template:
+        spec:
+        containers:
+        - image: "cfcontainerization/capi-release:opensuse-42.3-24.g63783b3-30.66-1.75.0"
+          command: ["some-ruby-script-that-renders"]
+          env:
+          - name: "INSTANCE_GROUP_NAME"
+            value: "instance-group-2"
+          volumeMounts:
+          - name: deployment-manifest
+            mountPath: /opt/fissile/
+          volumeMounts:
+          - name: mydeployment-bar
+            mountPath: /opt/fissile/
+        volumes:
+        - name: deployment-manifest
+          secret:
+            name: "deployment-mydeployment-12"
+            items:
+            - key: "manifest"
+              path: "deployment-manifest.yaml"
+        - name:
+          secret:
+            name: "mydeployment-bar"
+            items:
+            - key: "value"
+              path: "bar"
+```
+
+The following is an example output for `instance-group-2` and `release-a`
+```json
+{
+    "job1":{
```

Comment:
Any particular reason to have the release name as a key, instead of including it in the path name, like `job1/bin/start_ctl.sh`?

It provides the additional flexibility to mount the whole secret under the `/var/vcap/jobs` mount point without having to enumerate each file individually.

Reply:
Ok, further research shows that you cannot have slashes in key names; they need to match `[-._a-zA-Z0-9]+`.

This also means that the whole configmap cannot be mounted automatically and maintain the directory structure; that is only possible with PVs afaict.

We will have to specify the mount path for each file individually. On the plus side this means we also get to specify the file mode separately, so only files under `bin/` need to be made executable.

I would suggest that we simply number all rendered files, and then include an extra key that just concatenates the relative paths for each file, e.g.

```
{
    "1": "content of file 1",
    "2": "content of file 2",
    "_": "job1/bin/pre-start:job1/config/my.conf"
}
```

We could maybe use serialized JSON instead of joining the paths with `:`, but I don't care much either way.

Reply:
Actually, this may need more bike-shedding: it will probably be useful to have the mapping for each job separately. At which time using serialized JSON becomes more attractive. So instead of

```
    "_": "job1/bin/pre-start:job1/config/my.conf"
```

we would have

```
    "job1": "{ \"bin/prestart\": \"1\", "config/my.conf\": \"2\" }"
```

with additional entries for each additional job in the release.